projects/odf_transform/odf_transform/odf_parser/odf_parser.py

```diff
@@ -165,68 +179,119 @@ def _convert_to_number(value):
     return metadata, data_raw
 
 
-def define_odf_variable_attributes(metadata,
-                                   organizations=None,
-                                   vocabulary=None,
-                                   vocabulary_attribute_list=None,
-                                   odf_var_header_prefix='original_',
-                                   odf_variable_name="CODE",
-                                   flag_prefix='QualityFlag:'):
+def odf_flag_variables(metadata, flag_convention=None):
+    """
+    odf_flag_variables handle the different conventions used within the ODF files over the years and map them
+     to the CF standards.
+    """
+
+    # Loop through each variables and detect flag variables
+    previous_key = None
+    for var, att in metadata.items():
+        # Retrieve information from variable name
+        odf_var_name = parse_odf_code_variable(var)
+        related_variable = None
+
+        # FLAG VARIABLES Detect if it is a flag column
+        is_q_flag = var.startswith("Q") and var[:1] in metadata.keys()
+        is_qqqq_flag = odf_var_name['name'] == 'QQQQ'
+        is_general_flag = odf_var_name['name'] in ['QCFF', 'FFFF']
+
+        is_flag_column = is_q_flag or is_qqqq_flag or is_general_flag
+
+        # Find related variable
+        if is_qqqq_flag:
+            # MLI FLAG QQQQ should apply to previous variable
+            related_variable = previous_key
+        if is_q_flag:
+            # Q  Format is usually Q+[PCODE] of the associated variable
+            related_variable = var[1:]
+
+        # Set previous key for the next iteration
+        previous_key = var
+
+        # If the variable isn't a flag variable, go to the next iteration
+        if not is_flag_column:
+            continue
+
+        # If flag is specific to a variable, try to confirm if the odf name of the related variable match the
+        # flag name. If yes, add this flag variable to the ancillary_attribute of the related variable
+        if related_variable:
```

Comment:
alternatively you could do
```
if not related_variable:
     continue
```
otherwise your check below - 
`if flag_convention:` still runs even if there's no related_variable found

Reply:
or you could raise an error if `related_variable` not set

Reply:
so flag variables aren't related to any specific variables QCFF and FFFF. Not sure that would work with this.